addons/product/models/product.py

```diff
@@ -65,6 +65,11 @@ def unlink(self):
         main_category = self.env.ref('product.product_category_all')
         if main_category in self:
             raise UserError(_("You cannot delete this product category, it is the default generic category."))
+        external_ids = self.get_external_id()
+        for category in self:
+            external_id = external_ids[category.id]
+            if external_id and not external_id.startswith('__export__'):
+                raise UserError(_("You cannot delete the %s product category.", category.display_name))
```

Comment:
Why forbiding all referenced records deletion?

Reply:
Hello @Feyensv ,
Requested changes are done.
We are forbidding the deletion of referenced records because, if we delete them and then later when we try to install the module then it will  raise ‘Parserror’. And we are only forbidding the deletion of categories which are created through data files and not manually created. The user would be able to delete manually created records. So the primary purpose was to prevent the traceback.

Reply:
Yes, I understand what you're trying to do, but restricting the records deletion has been a long time debate, and every time, it ended with "we must avoid restricting the deletion as much as possible", so we try to limit the restriction only to the specific records that have been known to break existing flows/uninstall/install/...